### PR TITLE
Store non utf-8 characters properly in json files

### LINKED ISSUE #206: Warning: Error in gsub: input string 1 is invalid UTF-8
When recording a test with the shinytest package using recordTest(), everything works fine.
But when I rerun the test with the following statement:
testApp(".")
I get an error message:
Warning: Error in gsub: input string 1 is invalid UTF-8
The encoding of all shiny files (global.R, server.R and ui.R) is UTF-8. The ui + server files contain special characters.
I run the shinytest package 1.3.0 on a windows 10 pc with R version 3.5.0 and Rstudio version 1.1.453 and the test json file is:

{
  "input": {
    "action0": 0,
    "algemeen": "Herkomst",
    "bevolking": "Ja",
    "CMI": "Globaal",
    "CMIKeuze": "Alle",
    "jaren1": "2017",
    "jaren2": "2016",
    "jaren2T": "2016",
    "jaren3": "2017",
    "jaren3T": "2017",
    "jaren4": "2017",
    "jaren4T": "2017",
    "jaren5T": "2017",
    "jaren6": "2017",
    "jaren6T": "2017",
    "jaren7": "2017",
    "KenletterKeuze": "Alle",
    "kilometer": 5,
    "MA": "Globaal",
    "MID": "Globaal",
    "MIDVBL": "Globaal",
    "mzgJaren2": [
      2016,
      2017
    ],
    "NHPPD": "BB per Oper VTE",
    "niveau": "0",
    "show1": 0,
    "show18": 0,
    "show19": 0,
    "show2": 0,
    "show21": 0,
    "show3": 0,
    "show4": 0,
    "show40": 0,
    "show5": 0,
    "show6": 0,
    "show7": 0,
    "VBL": "Globaal",
    "ZR": "provincie"
  },
  "output": {
    "Contact2": "Uw contactpersoon is Test",
    "Disclaimer": "test",
    "herkomstTitel": "<h4 align=\"center\">**Herkomst patiënten**<\/h4>",
    "mymap": {
      "x": {
        "options": {
          "zoomControl": false
        }
      },
      "evals": [

      ],
      "jsHooks": [

      ],
      "deps": [

      ]
    },
    "semester1": {
      "html": "Opgelet: enkel 1ste semester!",
      "deps": [

      ]
    },
    "show": {
      "message": "",
      "call": "NULL",
      "type": [
        "shiny.silent.error",
        "validation"
      ]
    },
    "showdetails": {
      "message": "",
      "call": "NULL",
      "type": [
        "shiny.silent.error",
        "validation"
      ]
    },
    "tekstzkh": "test"
  },
  "export": {

  }
}

### PR DESCRIPTION
**Pull request by AlexandraKapp**
Fixes #206

### CODE REVIEW

File: R/snapshot.R [hunk 1]
```diff
@@ -51,7 +51,9 @@ sd_snapshot <- function(self, private, items, filename, screenshot)
   content <- raw_to_utf8(req$content)
   content <- hash_snapshot_image_data(content)
   content <- jsonlite::prettify(content, indent = 2)
-  writeChar(content, file.path(current_dir, filename), eos = NULL)
+  con <- file(file.path(current_dir, filename), encoding = "UTF-8")
+  write(content, con)
+  close(con)
```
Comment (wch) [inline R/snapshot.R:56]: It is possible for the `write` to throw an error, so the `close(con)` should be in an `on.exit()`, before `write()` is called.

**Comment by wch** [comment]
I had some comments, but now that I've thought about it a bit more, I think a better way to go is to just borrow the `writeUTF8` function from shiny and put it in shinytest:
https://github.com/rstudio/shiny/blob/e21a9a095e9a4f3f57aad479a7b62e4fd461aa32/R/utils.R#L1576-L1579

**Comment by AlexandraKapp** [comment]
> 
> 
> I had some comments, but now that I've thought about it a bit more, I think a better way to go is to just borrow the `writeUTF8` function from shiny and put it in shinytest:
> https://github.com/rstudio/shiny/blob/e21a9a095e9a4f3f57aad479a7b62e4fd461aa32/R/utils.R#L1576-L1579

thanks for your're feedback! I'll try it out

**Comment by AlexandraKapp** [comment]
> 
> 
> I had some comments, but now that I've thought about it a bit more, I think a better way to go is to just borrow the `writeUTF8` function from shiny and put it in shinytest:
> https://github.com/rstudio/shiny/blob/e21a9a095e9a4f3f57aad479a7b62e4fd461aa32/R/utils.R#L1576-L1579

Not 100% sure if I got this right, but I dont think it makes a big difference if `writeUT8` or `write` is used. as `content` is in proper utf8 format before it is stored. The error occurs when writing the file. So the encoding needs to be passed with the connection `con <- file(file.path(current_dir, filename), encoding = "UTF-8")` - both, in `write` or`writeLines` (whis is used in`writeUTF8`).

I'm not sure though, where to properly put the `close(con) ` in `on.exit()` as you mentioned.

**Comment by wch** [comment]
We've used the `writeUTF8` function in Shiny for a long time, so we're fairly confident that it works properly. For consistency and simplicity, it would be better to just bring that into utils.R, and then call it here.

If you just bring in that function and call it, then you don't have to worry about managing the connection manually.

**Comment by AlexandraKapp** [comment]
works 👌

**Comment by wch** [comment]
Thanks!

**Pull request merged**